cmd/buildkitd/main_containerd_worker.go

```diff
@@ -320,3 +367,13 @@ func validContainerdSocket(cfg config.ContainerdConfig) bool {
 	}
 	return true
 }
+
+// getRuntimeOptionsType gets empty runtime options by the runtime type name.
+func getRuntimeOptionsType(t string) interface{} {
+	switch t {
+	case plugin.RuntimeRuncV2:
+		return &runcoptions.Options{}
+	default:
+		return &runtimeoptions.Options{}
+	}
+}
```

Comment:
This is pulled pretty much directly from https://github.com/containerd/containerd/blob/c33249cbe6cb5a5e877647626472d16174b83f85/pkg/cri/sbserver/helpers.go#L289-L299.

Because of how this works, this means that we'd need to add support for each containerd runtime whose options we want to support, which is a little bit fiddly IMO.

@AkihiroSuda is there a different way to work around this? Or alternatively, could we potentially expose this logic as public in containerd?

Reply:
> Or alternatively, could we potentially expose this logic as public in containerd?

SGTM

Reply:
Ideally, options should be an opaque type. Otherwise, this is just a leaky abstraction and we only _pretend_ that runtimes are easily swappable.

Reply:
Just came across this (because it broke Windows `RUN` due to the aforementioned fiddly changes being needed, see #4364)

The only way I can see of avoiding this fiddlyness (which as-noted, containerd/cri also suffers from) would be to amend the containerd v2 runtime ABI to require runtimes to support both their own Options type, and accept `runtimeoptions.Options` and parse their own type out of either the `ConfigPath` or `ConfigBody`.

I suggested that hcsshim support that in https://github.com/microsoft/hcsshim/issues/1941 and have more notes on the idea there, but I'm not super-sold on it being a great approach. It also doesn't help a lot when an application needs to do more than reserialise a user-provided config block into an Options struct. But for use-cases like this (and containerd/cri) that's what we're trying to do, so it fits the bill well.

One improvement I did suggest is that the thing that defines the Options struct should _also_ export the name, so that if-chains like this are correctly coupled. But that assumes the struct is containerd-specific; it is in hcsshim, but I don't know off-hand if runc also uses this struct elsewhere and the runtime name would be irrelevant. (`plugin.RuntimeRuncV2` is historical legacy, AFAIK; the list of runtimes in `plugin` shouldn't be growing.)

I suspect that if the `default` option was `nil`, then at least it would not fail when a runtime whose options type is not known is used, as long as that runtime does not _require_ any settings; hcsshim meets this criteria, it correctly handles the "No options struct" case, it only barfs when it gets one it cannot deserialize.

In the end, making something like this public from containerd isn't a bad idea, but it only helps for cases that containerd knows about, and there'll certainly be situations where you need to handle a new runtime locally like this either because the containerd change is too new to vendor, or because the update to containerd is stuck in a PR.

Right now, containerd itself doesn't know about _any_ of these types; containerd/cri does because it does the same thing we do here (except it also populates `runtimeoptions.Options.ConfigBytes`) to handle its own config, and `ctr` happens to know how to pass "Debug" into hcsshim. And similar to my comment above, containerd probably _shouldn't_ start to know about all the possible runtime types, but no specifically-better solution comes to mind.